tests/update/patch_before_migration.pm

```diff
@@ -73,6 +79,53 @@ sub patching_sle {
     set_var("SYSTEM_PATCHED", 1);
 }
 
+sub install_package {
+    my @pk_list = split(/,/, get_var('PACKAGES'));
+    for my $pk (@pk_list) {
+        zypper_call "in -t package $pk";
+    }
+}
+
+sub install_patterns {
+
+    my $pcm = 0;
+    my @pt_list;
+    my @pt_list_un;
+    my @pt_list_in;
+
+    # install all patterns from product.
+    if (check_var('PATTERNS', 'all')) {
+        @pt_list_un = split(/ /, script_output("zypper pt -u | grep '^ ' | awk '{print \$2}' | sort -u | xargs"));
+        @pt_list_in = split(/ /, script_output("zypper pt -i | grep '^i' | awk -F '|' '{print \$2}' | sort -u | xargs"));
+        my %installed_pt = ();
+        foreach (@pt_list_in) {
+            $installed_pt{$_} = 1;
+        }
+        @pt_list = sort grep(!$installed_pt{$_}, @pt_list_un);
+        $pcm = grep /Amazon-Web-Services|Google-Cloud-Platform|Microsoft-Azure/, @pt_list_in;
+    }
+    #install certain pattern from parameter.
+    else {
+        @pt_list = split(/,/, get_var('PATTERNS'));
+    }
+
+    for my $pt (@pt_list) {
+        # Cloud patterns are conflict by each other, only install cloud pattern from single vender.
+        if ($pt =~ /Amazon-Web-Services|Google-Cloud-Platform|Microsoft-Azure/) {
+            if ($pcm == 0) {
+                zypper_call "in -t pattern $pt*";
```

Comment:
How about use zypper_call only once at later, and just check whether clould patterns installed here? Maybe: "next unless $pcm == 0" ?